{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolor\n",
    "import os\n",
    "import json\n",
    "from Bio import Phylo\n",
    "import numpy as np\n",
    "import gzip\n",
    "from sklearn.metrics import roc_auc_score\n",
    "import seaborn as sns\n",
    "import plotly.express as px\n",
    "from ete3 import Tree\n",
    "from scipy import stats\n",
    "from sklearn.decomposition import PCA\n",
    "import umap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "matplotlib.rcParams['font.family']       = 'Arial'\n",
    "matplotlib.rcParams['font.sans-serif']   = [\"Arial\",\"DejaVu Sans\",\"Lucida Grande\",\"Verdana\"]\n",
    "matplotlib.rcParams['figure.figsize']    = [4,3]\n",
    "matplotlib.rcParams['font.size']         = 10\n",
    "matplotlib.rcParams[\"axes.labelcolor\"]   = \"#000000\"\n",
    "matplotlib.rcParams[\"axes.linewidth\"]    = 1.0 \n",
    "matplotlib.rcParams[\"xtick.major.width\"] = 1.0\n",
    "matplotlib.rcParams[\"ytick.major.width\"] = 1.0\n",
    "cmap1 = plt.cm.tab20\n",
    "cmap2 = plt.cm.Set3  \n",
    "#plt.style.use('default')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/Users/konnonaoki/Documents/backupped/Research/IwasakiLab/Data/MetabolicNetworkEvolution/experiment/NK_M0153\")\n",
    "\n",
    "for dir in [\"figures\", \"tables\", \"networks\", \"itol\", \"tree\", \"itol/ko_ex_prevanc\", \"itol/evo_pred\"]:\n",
    "    try:\n",
    "        os.mkdir(dir)\n",
    "    except:\n",
    "        None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Classess of KOs\n",
    "\n",
    "table_dir = \"/Users/konnonaoki/Documents/backupped/Research/IwasakiLab/Data/MetabolicNetworkEvolution/experiment/NK_M0151\"\n",
    "\n",
    "df_path_ko = pd.read_table(table_dir + \"/tables/path_ko.txt\", names = ['Pathway', 'KO'])\n",
    "df_rn_ko = pd.read_table(table_dir + \"/tables/rn_ko.txt\", names = ['Reaction','KO'])\n",
    "df_md_ko = pd.read_table(table_dir + \"/tables/md_ko.txt\", names = ['Module','KO'])\n",
    "df_path_md = pd.read_table(table_dir + \"/tables/path_md.txt\", names = ['Pathway','Module'])\n",
    "ontology = json.load(open(\"/Users/konnonaoki/Documents/backupped/Research/IwasakiLab/Data/MetabolicNetworkEvolution/experiment/NK_M0151/json/ko00001.json\"))\n",
    "\n",
    "ontology_tree = Phylo.BaseTree.Tree(Phylo.BaseTree.Clade(name=ontology['name']))\n",
    "root_clade    = Phylo.BaseTree.Clade(name=ontology['name'])\n",
    "stack = [(ontology, root_clade)]\n",
    "\n",
    "while len(stack) > 0:\n",
    "    term, clade = stack.pop()\n",
    "    if ('children' in term.keys()):\n",
    "        for child in term['children']:\n",
    "            child_clade = Phylo.BaseTree.Clade(name = child['name'])\n",
    "            clade.clades.append(child_clade)\n",
    "            stack.append((child, child_clade))\n",
    "\n",
    "ontology_tree = Phylo.BaseTree.Tree(root_clade)\n",
    "\n",
    "list_category_ko = []\n",
    "for clade in ontology_tree.clade.clades[0].clades:\n",
    "    for tip in clade.get_terminals():\n",
    "        KO = tip.name.split()[0]\n",
    "        if (KO[0] == 'K'):\n",
    "            list_category_ko.append([clade.name, KO])\n",
    "df_category_ko = pd.DataFrame(list_category_ko, columns = ['category', 'KO'])\n",
    "st_category_ko = []\n",
    "for clade in ontology_tree.clade.clades[0].clades:\n",
    "    for tip in clade.get_terminals():\n",
    "        KO = tip.name.split()[0]\n",
    "        if (KO[0] == 'K'):\n",
    "            list_category_ko.append([clade.name, KO])\n",
    "df_category_ko = pd.DataFrame(list_category_ko, columns = ['category', 'KO'])\n",
    "df_category_ko = df_category_ko[~df_category_ko.duplicated()]\n",
    "\n",
    "df_ko_count = pd.DataFrame(df_category_ko.KO.value_counts())\n",
    "set_ko_with_unique_category = set(df_ko_count[df_ko_count['KO']==1].index)\n",
    "df_category_ko['unique'] = [(ko in set_ko_with_unique_category) for ko in df_category_ko.KO]\n",
    "df_uniquecategory_ko = df_category_ko[df_category_ko['unique']]\n",
    "\n",
    "# color of function categories\n",
    "\n",
    "colors = ['#66C2A5', '#FC8D62', '#8DA0CB', '#E78AC3', '#555555', '#FC8D62', '#8DA0CB', '#E78AC3', '#66C2A5', '#FC8D62', '#000000']\n",
    "\n",
    "cm_name = 'Set3' # B->G->R\n",
    "cm = plt.get_cmap(cm_name)\n",
    "\n",
    "df_category_ko_module = pd.merge(df_category_ko, df_md_ko, on = 'KO')\n",
    "df_category_ko_module['Nko'] = 1\n",
    "df_category_module_count = df_category_ko_module.groupby(['category', 'Module'], as_index = False).sum()\n",
    "df_maxcategory_module = df_category_module_count.loc[df_category_module_count.groupby('Module')['Nko'].idxmax(),:].sort_values('category')\n",
    "df_maxcategory_module = df_maxcategory_module.reset_index().loc[:, ['category', 'Module']]\n",
    "df_category_color = pd.DataFrame([[category, i] for i, category in enumerate(df_maxcategory_module.category.unique())], columns = [\"category\", 'category_id'])\n",
    "df_category_color['color'] = [mcolor.rgb2hex(cm(i)) for i in df_category_color['category_id']]\n",
    "#df_category_color\n",
    "\n",
    "df_category_ko_pathway = pd.merge(df_category_ko, df_path_ko, on = 'KO')\n",
    "df_category_ko_pathway['Nko'] = 1\n",
    "df_category_pathway_count = df_category_ko_pathway.groupby(['category', 'Pathway'], as_index = False).sum()\n",
    "df_maxcategory_pathway = df_category_pathway_count.loc[df_category_pathway_count.groupby('Pathway')['Nko'].idxmax(),:].sort_values('category')\n",
    "df_maxcategory_pathway = df_maxcategory_pathway.reset_index().loc[:, ['category', 'Pathway']]\n",
    "df_ko_desc = pd.read_table(table_dir + \"/tables/ko_desc.txt\", names = ['KO', 'Description'])\n",
    "df_uniquecategory_ko"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Open results of ongoing evolution prediction + reference phylogeny"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_future = pd.read_table(gzip.open(\"result/sp_prob.test.txt.gz\"), names = ['KO', 'target', 'pred_method', 'selec_method', 'opt_mode', 'Nfeatures', 'species', 'prob'])\n",
    "df_future"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all pairs of OG and extant species\n",
    "df_ko_sp_enumerated = df_future[(df_future['pred_method']=='LR') & (df_future['selec_method']=='ANOVA') & (df_future['opt_mode']=='N_opt_for_AUC_of_the_OG')].loc[:,['KO', 'species']].drop_duplicates() \n",
    "df_ko_sp_enumerated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = Phylo.read(\"/Users/konnonaoki/Documents/backupped/Research/IwasakiLab/Data/MetabolicNetworkEvolution/experiment/NK_M0150/tree/bac120_msa_r89.faa.mlgtdb.representative.renamed.rooted.nwk\", \"newick\")\n",
    "sp2ancsp = {}\n",
    "for internal in tree.get_nonterminals():\n",
    "    for child in internal.clades:\n",
    "        sp2ancsp[child.name] = internal.name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### possession ratio for each pair of KO and species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gn_ko_posratio = pd.read_table(\"../NK_M0145/list/sp_gn_posratio.txt\", names = [\"sptaxid\", \"KO\", \"posratio\"])\n",
    "df_Ngenomes = pd.read_table(\"/Users/konnonaoki/Documents/backupped/Research/IwasakiLab/Data/MetabolicNetworkEvolution/experiment/NK_M0145/list/analyzed_Ngenomes.txt\",names=[\"#genomes\",\"sptaxid\",\"kegg\",\"repgenome\",\"name\", '#genomes_up_to_100'])\n",
    "df_gn_ko_posratio = pd.merge(df_gn_ko_posratio,df_Ngenomes.loc[:,[\"sptaxid\",\"kegg\"]])\n",
    "df_gn_ko_posratio['species'] = [sp.split(':')[1] for sp in df_gn_ko_posratio['kegg']]\n",
    "df_ko_sp_posratio = df_gn_ko_posratio.loc[:, ['KO', 'species', 'posratio']]\n",
    "df_ko_sp_posratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(set(df_ko_sp_posratio['species']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree =  Tree(\"/Users/konnonaoki/Documents/backupped/Research/IwasakiLab/Data/MetabolicNetworkEvolution/experiment/NK_M0150/tree/bac120_msa_r89.faa.mlgtdb.representative.renamed.rooted.nwk\", format = 1)\n",
    "tree.prune(set(df_ko_sp_posratio['species']))\n",
    "open(\"tree/extracted_tree_114_species.nwk\", 'w').write(tree.write())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Presence/absence of each OFs in each species (ancestral and extant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read as a dataframe\n",
    "df_ko_sp_presence = pd.read_table(\"/Users/konnonaoki/Documents/backupped/Research/IwasakiLab/Data/MetabolicNetworkEvolution/experiment/NK_M0150/result/ko_gn_weight.mlgtdb_MPPA.txt\", names = ['KO', 'species', 'presence'])\n",
    "\n",
    "# read as a dictionary\n",
    "ko_sp2presence = {}\n",
    "for ko, sp, presence in zip(df_ko_sp_presence['KO'], df_ko_sp_presence['species'], df_ko_sp_presence['presence']):\n",
    "    if ko not in ko_sp2presence.keys():\n",
    "        ko_sp2presence[ko] = {}\n",
    "    ko_sp2presence[ko][sp] = presence\n",
    "\n",
    "df_ko_sp_presence = pd.merge(df_ko_sp_enumerated, df_ko_sp_presence, on = ['KO', 'species'], how = 'left')\n",
    "df_ko_sp_presence['presence'] = df_ko_sp_presence['presence'].fillna(0)\n",
    "    \n",
    "df_ko_sp_presence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ko_sp_presence['anc_species'] = [sp2ancsp[sp] if sp != 'Bacteria' else '' for sp in df_ko_sp_presence['species']]\n",
    "df_ko_sp_presence['anc_presence'] = [(ko_sp2presence[ko][anc] if anc in ko_sp2presence[ko].keys() else 0) if ko in ko_sp2presence.keys() else 0 for ko, anc in zip(df_ko_sp_presence['KO'], df_ko_sp_presence['anc_species'])]\n",
    "df_ko_sp_presence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyzed_sp_set = set(df_ko_sp_posratio['species'])\n",
    "df_ko_sp_presence['pangenome_analyzed'] = [(sp in analyzed_sp_set) for sp in df_ko_sp_presence['species']]\n",
    "df_ko_sp_presence_ext = df_ko_sp_presence[df_ko_sp_presence['pangenome_analyzed']]\n",
    "df_ko_sp_presence_ext"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Presence-AbsenceのiTOL label \n",
    "for KO in list(set(df_ko_sp_presence['KO'])):\n",
    "\n",
    "    df_ko_sp_presence_ko = df_ko_sp_presence[df_ko_sp_presence['KO']==KO]\n",
    "\n",
    "    color_absent = \"#E5E5E5\"\n",
    "    color_absent_newly = \"#CCCCCC\"\n",
    "    \n",
    "    color_present = \"#3E8D27\"\n",
    "    color_present_newly = \"#97F788\"\n",
    "\n",
    "    with open(\"itol/ko_ex_prevanc/annotation.ex_prevanc.\"+KO+\".txt\", 'w') as handle:\n",
    "        handle.write(\n",
    "            \"DATASET_COLORSTRIP\\nSEPARATOR SPACE\\nDATASET_LABEL \"+KO+\"\\nCOLOR #ff0000\\nDATA\\n\"\n",
    "        )\n",
    "\n",
    "        for species, presence, anc_presence in zip(df_ko_sp_presence_ko['species'], df_ko_sp_presence_ko['presence'], df_ko_sp_presence_ko['anc_presence']):\n",
    "            if(abs(presence - 1) < 0.01 and anc_presence > 0.9):\n",
    "                handle.write(\n",
    "                    species + \" \" + color_present + \" present_\"+str(anc_presence)+\"\\n\"\n",
    "                )\n",
    "            elif(abs(presence - 1)<0.01 and anc_presence < 0.9):\n",
    "                handle.write(\n",
    "                    species + \" \" + color_present_newly + \" newly_present_\"+str(anc_presence)+\"\\n\"\n",
    "                )\n",
    "            elif(abs(presence - 0)<0.01 and anc_presence > 0.1):\n",
    "                handle.write(\n",
    "                    species + \" \" + color_absent_newly + \" newly_absent_\"+str(anc_presence)+\"\\n\"\n",
    "                )\n",
    "            elif(abs(presence - 0)<0.01 and anc_presence < 0.1):\n",
    "                handle.write(\n",
    "                    species + \" \" + color_absent + \" absent_\"+str(anc_presence)+\"\\n\"\n",
    "                )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Combine dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ko_sp_presence_ext_future = pd.merge(df_ko_sp_presence_ext, df_future, on = ['KO', 'species'], how = 'left')\n",
    "df_ko_sp_presence_ext_future"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ko_sp_presence_ext_future_posratio = pd.merge(df_ko_sp_presence_ext_future, df_ko_sp_posratio, on = ['KO', 'species'], how = 'left')\n",
    "df_ko_sp_presence_ext_future_posratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ko_sp_presence_ext_future_posratio['posratio'] = df_ko_sp_presence_ext_future_posratio['posratio'].fillna(0)\n",
    "df_ko_sp_presence_ext_future_posratio"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Histogram of gene possession ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_presence_posratio = \\\n",
    "    df_ko_sp_presence_ext_future_posratio[\n",
    "        #(df_ko_sp_presence_ext_future_posratio['target']=='gain') &\n",
    "        (df_ko_sp_presence_ext_future_posratio['pred_method']=='LR') &\n",
    "        (df_ko_sp_presence_ext_future_posratio['selec_method']=='ANOVA') &\n",
    "        (df_ko_sp_presence_ext_future_posratio['opt_mode']=='N_opt_for_AUC_of_the_OG') \n",
    "    ].loc[:,['KO', 'species', 'presence', 'posratio']]\n",
    "\n",
    "df_presence_posratio_category = pd.merge(df_presence_posratio, df_uniquecategory_ko, on = 'KO').drop_duplicates()\n",
    "df_presence_posratio_category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Nnot_possessing_in_KEGG_but_possessed_by_some_strain = len(df_presence_posratio_category[(df_presence_posratio_category[\"presence\"]==0) & (df_presence_posratio_category[\"posratio\"]>0)])\n",
    "Nnot_possessing_in_KEGG_but_possessed_by_some_strain / 239058"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "for category in list(set(df_presence_posratio_category['category']))+['any']:\n",
    "    \n",
    "    #if(True):\n",
    "    #category = '09105 Amino acid metabolism'\n",
    "    \n",
    "    if (category =='any'):\n",
    "        df_presence_posratio_category_ext = df_presence_posratio_category\n",
    "    else:    \n",
    "        df_presence_posratio_category_ext = df_presence_posratio_category[df_presence_posratio_category['category'] == category]\n",
    "\n",
    "    fig = plt.figure(figsize=(2.5,1.7))\n",
    "    ax = fig.add_axes([0.1,0.1,0.8,0.8])\n",
    "    ax.hist(\n",
    "        [\n",
    "            df_presence_posratio_category_ext[(df_presence_posratio_category_ext['posratio']>0) & ((df_presence_posratio_category_ext['presence']>0)==1)]['posratio'], \n",
    "            df_presence_posratio_category_ext[(df_presence_posratio_category_ext['posratio']>0) & ((df_presence_posratio_category_ext['presence']>0)==0)]['posratio']\n",
    "        ], bins=40, color=['#75868C', \"#DDE2E0\"],stacked=True\n",
    "    )\n",
    "\n",
    "    #ax.set_yscale(\"log\")\n",
    "    plt.gca().spines['right'].set_visible(False)\n",
    "    plt.gca().spines['top'].set_visible(False)\n",
    "    ax.set_xlabel(\"Possesion ratio\")\n",
    "    ax.set_title(category)\n",
    "    \n",
    "    ax2 = fig.add_axes([-0.12,0.1,0.03,0.8])\n",
    "    ax2.hist(\n",
    "        [\n",
    "            df_presence_posratio_category_ext[(df_presence_posratio_category_ext['posratio']==0) & ((df_presence_posratio_category_ext['presence']>0)==1)]['posratio'], \n",
    "            df_presence_posratio_category_ext[(df_presence_posratio_category_ext['posratio']==0) & ((df_presence_posratio_category_ext['presence']>0)==0)]['posratio']\n",
    "        ], bins=1, color=['#75868C', \"#DDE2E0\"],stacked=True\n",
    "    )\n",
    "    \n",
    "    plt.gca().spines['right'].set_visible(False)\n",
    "    plt.gca().spines['top'].set_visible(False)\n",
    "    ax2.set_ylabel(\"# of (OG, Species)\")\n",
    "    ax2.set_xticks([0])\n",
    "    \n",
    "    plt.savefig(\"figures/posratio_distribution.\"+category+\".pdf\",bbox_inches='tight')\n",
    "    #plt.show()  \n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Visualize ongoing gene gain/loss prediction by scatter plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = 'gain'\n",
    "\n",
    "threshold = 0.10\n",
    "\n",
    "df_ko_sp_presence_ext_future_posratio\n",
    "\n",
    "df_presence_posratio = \\\n",
    "    df_ko_sp_presence_ext_future_posratio[\n",
    "        (df_ko_sp_presence_ext_future_posratio['target']==target) &\n",
    "        (df_ko_sp_presence_ext_future_posratio['pred_method']=='LR') &\n",
    "        (df_ko_sp_presence_ext_future_posratio['selec_method']=='ANOVA') &\n",
    "        (df_ko_sp_presence_ext_future_posratio['opt_mode']=='N_opt_for_AUC_of_the_OG') \n",
    "    ].loc[:,['KO', 'species', 'presence', 'anc_presence', 'posratio', 'prob']]\n",
    "\n",
    "df_presence_posratio_to_be_analyzed = df_presence_posratio[(df_presence_posratio['presence']==0) & (df_presence_posratio['anc_presence']==0)]\n",
    "\n",
    "gaining_nongaining_list = []\n",
    "\n",
    "all_count = 0\n",
    "count = 0\n",
    "\n",
    "for ko in list(set(df_presence_posratio_to_be_analyzed['KO'])):\n",
    "    \n",
    "    all_count += 1\n",
    "\n",
    "    df_presence_posratio_to_be_analyzed_ko = df_presence_posratio_to_be_analyzed[df_presence_posratio_to_be_analyzed['KO'] == ko]\n",
    "    \n",
    "    df_doing = df_presence_posratio_to_be_analyzed_ko[df_presence_posratio_to_be_analyzed_ko['posratio'] >= threshold]\n",
    "    df_nondoing = df_presence_posratio_to_be_analyzed_ko[df_presence_posratio_to_be_analyzed_ko['posratio'] < threshold]\n",
    "\n",
    "    if (len(df_doing) > 1 and len(df_nondoing) > 1): \n",
    "    \n",
    "        gaining_nongaining_list.append(\n",
    "            [\n",
    "                ko,\n",
    "                np.median(df_doing.prob),\n",
    "                np.median(df_nondoing.prob)\n",
    "            ]\n",
    "        )\n",
    "    \n",
    "    else:\n",
    "        \n",
    "        count += 1\n",
    "    \n",
    "print(\"Omitted \"+str(count)+\" OGs in \"+str(all_count)+\" OGs\")\n",
    "df_pred_result = pd.DataFrame(gaining_nongaining_list, columns = ['KO', 'doing_medprob', 'nondoing_medprob'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(3,3))\n",
    "ax = fig.add_axes([0.1,0.1,0.8,0.8])\n",
    "x = np.arange(0, 10)\n",
    "y = x\n",
    "ax.plot(x, y, color = \"#AAAAAA\", alpha = 0.5)\n",
    "\n",
    "ax.set_xlabel(\"Median gain prob. predicted\\nfor gaining species\")#,color='#C0C0C0')\n",
    "ax.set_ylabel(\"Median gain prob. predicted\\nfor non-gaining species\")#,color='#544DE6')\n",
    "\n",
    "min_lim = 0.0001\n",
    "df_pred_result['doing_medprob'] = [max(prob, 0.0001) for prob in df_pred_result['doing_medprob']]\n",
    "df_pred_result['nondoing_medprob'] = [max(prob, 0.0001) for prob in df_pred_result['nondoing_medprob']]\n",
    "ax.scatter(df_pred_result[\"doing_medprob\"], df_pred_result[\"nondoing_medprob\"],alpha=0.5, s = 5, color =\"#645FCA\")#color=df_testresult['color_0to50vs50to100'])\n",
    "\n",
    "ax.set_xlim(min_lim,1.1)\n",
    "ax.set_ylim(min_lim,1.1)\n",
    "ax.set_xscale('log')\n",
    "ax.set_yscale('log')\n",
    "plt.gca().spines['right'].set_visible(False)\n",
    "plt.gca().spines['top'].set_visible(False)\n",
    "\n",
    "plt.savefig(\"figures/gaining.any.pdf\",bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = 'loss'\n",
    "\n",
    "gaining_nongaining_list = []\n",
    "\n",
    "for threshold in [0.90, 0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98, 0.99]:\n",
    "    df_presence_posratio = \\\n",
    "        df_ko_sp_presence_ext_future_posratio[\n",
    "            (df_ko_sp_presence_ext_future_posratio['target']==target) &\n",
    "            (df_ko_sp_presence_ext_future_posratio['pred_method']=='LR') &\n",
    "            (df_ko_sp_presence_ext_future_posratio['selec_method']=='ANOVA') &\n",
    "            (df_ko_sp_presence_ext_future_posratio['opt_mode']=='N_opt_for_AUC_of_the_OG') \n",
    "        ].loc[:,['KO', 'species', 'presence', 'anc_presence', 'posratio', 'prob']]\n",
    "\n",
    "    df_presence_posratio_to_be_analyzed = df_presence_posratio[(df_presence_posratio['presence']==1) & (df_presence_posratio['anc_presence']==1)]\n",
    "\n",
    "    all_count = 0\n",
    "    count = 0\n",
    "\n",
    "    for ko in list(set(df_presence_posratio_to_be_analyzed['KO'])):\n",
    "\n",
    "        all_count += 1\n",
    "\n",
    "        df_presence_posratio_to_be_analyzed_ko = df_presence_posratio_to_be_analyzed[df_presence_posratio_to_be_analyzed['KO'] == ko].reset_index()\n",
    "\n",
    "        df_doing = df_presence_posratio_to_be_analyzed_ko[df_presence_posratio_to_be_analyzed_ko['posratio'] < threshold]\n",
    "        df_nondoing = df_presence_posratio_to_be_analyzed_ko[df_presence_posratio_to_be_analyzed_ko['posratio'] >= threshold]\n",
    "        df_presence_posratio_to_be_analyzed_ko[\"doing\"] = (df_presence_posratio_to_be_analyzed_ko['posratio'] < threshold)\n",
    "\n",
    "        if (len(df_doing) > 1 and len(df_nondoing) > 1): \n",
    "\n",
    "            gaining_nongaining_list.append(\n",
    "                [\n",
    "                    threshold,\n",
    "                    ko,\n",
    "                    np.median(df_doing.prob),\n",
    "                    np.median(df_nondoing.prob),\n",
    "                    roc_auc_score(df_presence_posratio_to_be_analyzed_ko[\"doing\"], df_presence_posratio_to_be_analyzed_ko[\"prob\"])\n",
    "                ]\n",
    "            )\n",
    "\n",
    "        else:\n",
    "\n",
    "            count += 1\n",
    "\n",
    "    print(\"Omitted \"+str(count)+\" OGs in \"+str(all_count)+\" OGs\")\n",
    "\n",
    "df_pred_result = pd.DataFrame(gaining_nongaining_list, columns = ['threshold', 'KO', 'doing_medprob', 'nondoing_medprob', 'auc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pred_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold = 0.90\n",
    "df_pred_result_ext = df_pred_result[df_pred_result['threshold'] == threshold]\n",
    "\n",
    "fig = plt.figure(figsize=(3,3))\n",
    "ax = fig.add_axes([0.1,0.1,0.8,0.8])\n",
    "x = np.arange(0, 10)\n",
    "y = x\n",
    "ax.plot(x, y, color = \"#AAAAAA\", alpha = 0.5)\n",
    "\n",
    "ax.set_xlabel(\"Median loss prob. predicted\\nfor gaining species\")#,color='#C0C0C0')\n",
    "ax.set_ylabel(\"Median loss prob. predicted\\nfor non-gaining species\")#,color='#544DE6')\n",
    "\n",
    "min_lim = 0.0001\n",
    "df_pred_result_ext['doing_medprob'] = [max(prob, 0.0001) for prob in df_pred_result_ext['doing_medprob']]\n",
    "df_pred_result_ext['nondoing_medprob'] = [max(prob, 0.0001) for prob in df_pred_result_ext['nondoing_medprob']]\n",
    "ax.scatter(df_pred_result_ext[\"doing_medprob\"], df_pred_result_ext[\"nondoing_medprob\"],alpha=0.5, s = 5, color ='#E1BB63')#color=df_testresult['color_0to50vs50to100'])\n",
    "\n",
    "ax.set_xlim(min_lim,1.1)\n",
    "ax.set_ylim(min_lim,1.1)\n",
    "ax.set_xscale('log')\n",
    "ax.set_yscale('log')\n",
    "plt.gca().spines['right'].set_visible(False)\n",
    "plt.gca().spines['top'].set_visible(False)\n",
    "\n",
    "plt.savefig(\"figures/losing.any.pdf\",bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Draw histogram of AUCs by function, prediction method, and threshold"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Time-consuming! calculate AUC for each threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pred_result_list = []\n",
    "\n",
    "for pred_method in ['LR', 'RF']:\n",
    "    for selec_method in ['ANOVA', 'RandomForest']:\n",
    "        for opt_mode in ['N_opt_for_AUC_of_the_OG', 'N_opt_for_median_AUC']:\n",
    "\n",
    "            for target, threshold_list in [('gain', [0, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1]), ('loss', [0.89, 0.90, 0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98, 0.99])]:\n",
    "                \n",
    "                for threshold in threshold_list:\n",
    "                    \n",
    "                    df_presence_posratio = \\\n",
    "                        df_ko_sp_presence_ext_future_posratio[\n",
    "                            (df_ko_sp_presence_ext_future_posratio['target']==target) &\n",
    "                            (df_ko_sp_presence_ext_future_posratio['pred_method']==pred_method) &\n",
    "                            (df_ko_sp_presence_ext_future_posratio['selec_method']==selec_method) &\n",
    "                            (df_ko_sp_presence_ext_future_posratio['opt_mode']==opt_mode) \n",
    "                        ].loc[:,['KO', 'species', 'presence', 'anc_presence', 'posratio', 'prob']]\n",
    "\n",
    "                    \n",
    "                    #########Difference between 'gain' and 'loss'#######\n",
    "                    if target == 'gain':\n",
    "                        df_presence_posratio_to_be_analyzed = df_presence_posratio[(df_presence_posratio['presence']==0) & (df_presence_posratio['anc_presence']==0)]\n",
    "                    elif target == 'loss':\n",
    "                        df_presence_posratio_to_be_analyzed = df_presence_posratio[(df_presence_posratio['presence']==1) & (df_presence_posratio['anc_presence']==1)]\n",
    "                    #########################################\n",
    "                        \n",
    "                        \n",
    "                    all_count = 0\n",
    "                    count = 0\n",
    "\n",
    "                    for ko in list(set(df_presence_posratio_to_be_analyzed['KO'])):\n",
    "\n",
    "                        all_count += 1\n",
    "\n",
    "                        df_presence_posratio_to_be_analyzed_ko = df_presence_posratio_to_be_analyzed[df_presence_posratio_to_be_analyzed['KO'] == ko].reset_index()\n",
    "\n",
    "                        #########Difference between 'gain' and 'loss'#######\n",
    "                        if target == 'gain':\n",
    "                            df_presence_posratio_to_be_analyzed_ko[\"doing\"] = (df_presence_posratio_to_be_analyzed_ko['posratio'] > threshold)\n",
    "                        elif target == 'loss':\n",
    "                            df_presence_posratio_to_be_analyzed_ko[\"doing\"] = (df_presence_posratio_to_be_analyzed_ko['posratio'] < threshold)\n",
    "                        #########################################\n",
    "\n",
    "                        df_doing        = df_presence_posratio_to_be_analyzed_ko[df_presence_posratio_to_be_analyzed_ko[\"doing\"]]\n",
    "                        df_nondoing  = df_presence_posratio_to_be_analyzed_ko[~(df_presence_posratio_to_be_analyzed_ko[\"doing\"])]\n",
    "                        \n",
    "                        Ndoing = len(df_doing)\n",
    "                        \n",
    "\n",
    "                        if (len(df_doing) > 1 and len(df_nondoing) > 1):\n",
    "\n",
    "                            pred_result_list.append(\n",
    "                                [\n",
    "                                    target,\n",
    "                                    pred_method,\n",
    "                                    selec_method,\n",
    "                                    opt_mode,\n",
    "                                    threshold,\n",
    "                                    ko,\n",
    "                                    Ndoing,\n",
    "                                    np.median(df_doing.prob),\n",
    "                                    np.median(df_nondoing.prob),\n",
    "                                    roc_auc_score(df_presence_posratio_to_be_analyzed_ko[\"doing\"], df_presence_posratio_to_be_analyzed_ko[\"prob\"])\n",
    "                                ]\n",
    "                            )\n",
    "\n",
    "                        else:\n",
    "\n",
    "                            count += 1\n",
    "\n",
    "                    print(\"Omitted \"+str(count)+\" OGs in \"+str(all_count)+\" OGs\")\n",
    "\n",
    "df_pred_result = pd.DataFrame(pred_result_list, columns = ['target','pred_method','selec_method','opt_mode','threshold', 'KO', 'Ndoing', 'doing_medprob', 'nondoing_medprob', 'auc'])\n",
    "df_pred_result.to_csv('tables/pred_result.txt',sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pred_result = pd.read_csv('tables/pred_result.txt',sep = '\\t')\n",
    "df_pred_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for pred_method in ['LR', 'RF']:\n",
    "    for selec_method in ['ANOVA', 'RandomForest']:\n",
    "        for opt_mode in ['N_opt_for_AUC_of_the_OG', 'N_opt_for_median_AUC']:\n",
    "\n",
    "            for target, color  in [('gain',\"#645FCA\"),('loss',  '#E1BB63')]:\n",
    "                    \n",
    "                    df_pred_result_ext = \\\n",
    "                        df_pred_result[\n",
    "                            (df_pred_result['target']==target) &\n",
    "                            (df_pred_result['pred_method']==pred_method) &\n",
    "                            (df_pred_result['selec_method']==selec_method) &\n",
    "                            (df_pred_result['opt_mode']==opt_mode) \n",
    "                    ]\n",
    "                    \n",
    "\n",
    "                    fig = plt.figure(figsize=(4,1.5))\n",
    "                    ax = fig.add_axes([0.1,0.1,0.8,0.8])\n",
    "                    #sns.violinplot(data = df_pred_result, y = 'auc', linewidth=0.5, color = '#FFFFFF', alpha = 0)\n",
    "                    #sns.stripplot (data = df_pred_result, y = 'auc', linewidth=0,    size = 1, alpha=.5,jitter=0.3, dodge=True, palette=[color])\n",
    "\n",
    "                    sns.violinplot(data = df_pred_result_ext, x = 'threshold', y = 'auc', linewidth=0.5, color = '#FFFFFF', alpha = 0)\n",
    "                    sns.stripplot (data = df_pred_result_ext, x = 'threshold', y = 'auc', linewidth=0, size = 1, alpha=.5,jitter=0.3, dodge=True, palette=[color])\n",
    "                    ax.tick_params(axis='x', labelrotation= 90)\n",
    "                    ax.set_title(pred_method+\" \"+selec_method+\" \"+opt_mode+\" \"+target)\n",
    "                    #ax.get_legend().remove()\n",
    "                    ax.set_xlabel('Threshold')\n",
    "                    ax.set_ylabel('AUC')\n",
    "                    \n",
    "                    ax.set_ylim(-0.05, 1.05)\n",
    "                    \n",
    "                    plt.savefig(\"figures/threshold_AUC.\"+pred_method+\".\"+selec_method+\".\"+opt_mode+\".\"+target+\".pdf\", bbox_inches = 'tight')\n",
    "                    #plt.show()\n",
    "                    plt.close()\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pred_result_category = pd.merge(df_pred_result, df_uniquecategory_ko, on = 'KO')\n",
    "\n",
    "for pred_method in ['LR', 'RF']:\n",
    "    for selec_method in ['ANOVA', 'RandomForest']:\n",
    "        for opt_mode in ['N_opt_for_AUC_of_the_OG', 'N_opt_for_median_AUC']:\n",
    "\n",
    "            for target, color, threshold  in [('gain',\"#645FCA\", 0.05),('loss',  '#E1BB63', 0.90)]:\n",
    "                    \n",
    "                    df_pred_result_ext = \\\n",
    "                        df_pred_result_category[\n",
    "                            (df_pred_result_category['target']==target) &\n",
    "                            (df_pred_result_category['pred_method']==pred_method) &\n",
    "                            (df_pred_result_category['selec_method']==selec_method) &\n",
    "                            (df_pred_result_category['opt_mode']==opt_mode) &\n",
    "                            (df_pred_result_category['threshold']==threshold) \n",
    "\n",
    "                    ]\n",
    "                    \n",
    "                    category_order = list(reversed(list(df_pred_result_category[\n",
    "                            (df_pred_result_category['target']==\"gain\") &\n",
    "                            (df_pred_result_category['pred_method']==pred_method) &\n",
    "                            (df_pred_result_category['selec_method']==selec_method) &\n",
    "                            (df_pred_result_category['opt_mode']==opt_mode) &\n",
    "                            (df_pred_result_category['threshold']==0.05) ].groupby(\"category\", as_index=False).median().sort_values(\"auc\")[\"category\"])))\n",
    "                    \n",
    "                    fig = plt.figure(figsize=(1.8,3))\n",
    "                    ax = fig.add_axes([0.1,0.1,0.8,0.8])\n",
    "                    #sns.violinplot(data = df_pred_result, y = 'auc', linewidth=0.5, color = '#FFFFFF', alpha = 0)\n",
    "                    #sns.stripplot (data = df_pred_result, y = 'auc', linewidth=0,    size = 1, alpha=.5,jitter=0.3, dodge=True, palette=[color])\n",
    "\n",
    "                    sns.boxplot(data = df_pred_result_ext, y = 'category', x = 'auc', order = category_order, orient = \"h\", linewidth=0.5, color = '#FFFFFF',sym=\"\") # does not show outliers\n",
    "                    #sns.violinplot(data = df_pred_result_ext, x = 'category', y = 'auc', order = sorted(list(set(df_pred_result_ext['category']))), linewidth=0.5, color = '#FFFFFF', alpha = 0)\n",
    "                    sns.stripplot (data = df_pred_result_ext, y = 'category', x = 'auc', order = category_order,orient = \"h\",  linewidth=0, size = 2, alpha=.5,jitter=0.3, dodge=True, palette=[color])\n",
    "                    ax.set_title(pred_method+\" \"+selec_method+\" \"+opt_mode+\" \"+target)\n",
    "                    #ax.get_legend().remove()\n",
    "                    ax.set_ylabel('Category')\n",
    "                    ax.set_xlabel('AUC')\n",
    "                    \n",
    "                    ax.set_xlim(-0.05, 1.05)\n",
    "                    \n",
    "                    plt.savefig(\"figures/category_AUC.\"+pred_method+\".\"+selec_method+\".\"+opt_mode+\".\"+target+\".pdf\", bbox_inches = 'tight')\n",
    "                    #plt.show()\n",
    "                    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Gene gain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pred_method = \"RF\"\n",
    "selec_method = \"ANOVA\"\n",
    "opt_mode = 'N_opt_for_AUC_of_the_OG'\n",
    "\n",
    "#target = \"gain\"\n",
    "#threshold = 0.05\n",
    "\n",
    "for target, threshold, color in [(\"gain\", 0, \"#645FCA\"), (\"loss\", 0.90, '#E1BB63')]:\n",
    "\n",
    "    df_pred_result_ext = \\\n",
    "        df_pred_result[\n",
    "            (df_pred_result['target']==target) &\n",
    "            (df_pred_result['pred_method']==pred_method) &\n",
    "            (df_pred_result['selec_method']==selec_method) &\n",
    "            (df_pred_result['opt_mode']==opt_mode) &\n",
    "            (df_pred_result['threshold']==threshold)\n",
    "        ]\n",
    "    \n",
    "    fig = plt.figure(figsize=(2.3,1.2))\n",
    "    ax = fig.add_axes([0.1,0.1,0.8,0.8])\n",
    "    plt.gca().spines['right'].set_visible(False)\n",
    "    plt.gca().spines['top'].set_visible(False)\n",
    "    plt.gca().spines['bottom'].set_visible(False)\n",
    "    ax2 = fig.add_axes([0.1,-0.1,0.8,0.18])\n",
    "    plt.gca().spines['right'].set_visible(False)\n",
    "    plt.gca().spines['top'].set_visible(False)\n",
    "    plt.gca().spines['left'].set_visible(False)\n",
    "    ax.tick_params(bottom = False, labelbottom = False)\n",
    "    ax2.tick_params(left = False,)\n",
    "\n",
    "    sns.kdeplot(df_pred_result_ext[\"auc\"],color = color, ax = ax, fill=True)\n",
    "    sns.stripplot(data = df_pred_result_ext, x = \"auc\", color = color, size = 1, alpha=.5,jitter=0.4,)\n",
    "\n",
    "    ax.set_xlim(-0.1,1.1)\n",
    "    ax2.set_xlim(-0.1,1.1)\n",
    "    ax2.set_xlabel(\"AUC of \"+target+\" prediction\")\n",
    "\n",
    "    ax.set_title(target)\n",
    "    ax.set_ylim(0,3)\n",
    "    \n",
    "    ax.axvline(0.5, alpha = 0.5, lw = 0.75, color = \"#000000\")\n",
    "\n",
    "    plt.savefig(\"figures/kdeplot_AUC.\"+pred_method+\".\"+selec_method+\".\"+opt_mode+\".\"+target+\".pdf\", bbox_inches = 'tight')\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "    print(target, threshold, stats.ttest_1samp(df_pred_result_ext[\"auc\"], 0.5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for iTOL\n",
    "\n",
    "#target = 'gain'\n",
    "#threshold = 0.05\n",
    "#KO_list = ['K05898', 'K16842', 'K01214', 'K02805', 'K18824']\n",
    "\n",
    "for target, threshold, KO_list in [('gain', 0.05, ['K05898', 'K16842', 'K01214', 'K02805', 'K18824','K01601']), ('loss', 0.90, ['K01523', 'K19270', 'K01664', 'K00336', 'K00991'])]:\n",
    "\n",
    "    for KO in KO_list:\n",
    "\n",
    "        target_df = df_ko_sp_presence_ext_future_posratio[\n",
    "            (df_ko_sp_presence_ext_future_posratio['target']==target) &\n",
    "            (df_ko_sp_presence_ext_future_posratio['KO']==KO) &\n",
    "            (df_ko_sp_presence_ext_future_posratio['pred_method']=='RF') &\n",
    "            (df_ko_sp_presence_ext_future_posratio['selec_method']=='ANOVA') &\n",
    "            (df_ko_sp_presence_ext_future_posratio['opt_mode']=='N_opt_for_AUC_of_the_OG') \n",
    "        ]\n",
    "        \n",
    "        if target == 'gain':\n",
    "            target_df = target_df[(target_df['presence']==0)&(target_df['anc_presence']==0)]\n",
    "            target_df = target_df.reset_index()\n",
    "            target_df['doing'] = target_df['posratio'] > threshold\n",
    "        elif target =='loss':\n",
    "            target_df = target_df[(target_df['presence']==1)&(target_df['anc_presence']==1)]\n",
    "            target_df = target_df.reset_index()\n",
    "            target_df['doing'] = target_df['posratio'] < threshold\n",
    "\n",
    "        with open(\"itol/evo_pred/bar_prob.\"+KO+\".\"+pred_method+\".\"+selec_method+\".\"+opt_mode+\".\"+target+\".txt\", 'w') as handle:\n",
    "                handle.write(\n",
    "                    \"DATASET_MULTIBAR\\nSEPARATOR TAB\\nDATASET_LABEL\t\"+KO+\" \"+target+\" prob\\nCOLOR\t#A0A5EC\\nFIELD_COLORS\\t#8682E7\\t#AAAAAA\\nFIELD_LABELS\\tf1\\tf2\\nALIGN_FIELDS\\t0\\nMARGIN\t0\\nWIDTH\t100\\nDATA\\n\"\n",
    "                    )\n",
    "                for i in range(len(target_df)):\n",
    "                    name     = list(target_df[\"species\"])[i]\n",
    "                    prob = list(target_df[\"prob\"])[i]\n",
    "                    doing  = list(target_df[\"doing\"])[i]\n",
    "                    #posratio = target_df[\"posratio\"][i]\n",
    "                    if   doing:\n",
    "                        handle.write(name+\"\\t\"+str(prob)+\"\\t0\\n\")\n",
    "                    else:\n",
    "                        handle.write(name+\"\\t0\\t\"+str(prob)+\"\\n\")"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "6e735269e152434eedfcfab596c1b5bd92dbf01a7c5f35c3224738c65ab1c2b6"
  },
  "kernelspec": {
   "display_name": "Python 3.9.5 64-bit ('metabo': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
